### DISCUSSION TITLE
Does MS have automated UI tests for .NET Maui? If so, what framework do they use?

### DISCUSSION THREAD

**Discussion by JohnHDev** [dotnet/maui#6192] [Q&A] [answered]
We are looking at what would be required to migrate from XF projects to Maui and we currently have a dependency on Xamarin.UITest. 
Unfortunately Xamarin.UITest is not compatible with .NET 6 Maui projects for iOS, throwing: 

    Package Xamarin.TestCloud.Agent 0.23.1 is not compatible with net6.0-ios15.2

Xamarin UITest used to work up to around preview 12 of Maui, but broke after that.

What is Microsoft using for its automated UI tests?

**Comment by cdefgah** [comment]
I've found this video: https://www.youtube.com/watch?v=9LxQwpjKxhE

But package **Microsoft.Maui.Automation.Interactive** is not available on nuget.org feed.

![image](https://user-images.githubusercontent.com/10169696/195091531-2eb4e5d8-731b-477c-9a3e-c6dfeebdcc36.png)

**Reply by JohnHSE** [reply]
That nuget package was hosted locally and only provides some helpers for the demo I believe.

Also, his demo only covered android, which Xamarin UITest still works with. It is iOS that doesn't.

**Reply by cdefgah** [reply]
I just forgot to put "Include prerelease". When this flag is enabled, the package is being shown.

Check this: https://github.com/Redth/Maui.UITesting

Although the driver hangs up when I try to get an element by automation id :)

**Reply by JohnHSE** [reply]
oooo now *THAT* looks interesting....

**Comment by hartez** [comment, marked as answer]
> What is Microsoft using for its automated UI tests?

We're using [xharness](https://github.com/dotnet/xharness) and some [custom test runner code](https://github.com/dotnet/maui/tree/main/src/TestUtils) for on-device testing. You can look at the various DeviceTests projects scattered throughout the repo (e.g., [the Core device [tests](https://github.com/dotnet/maui/tree/main/src/Core/tests/DeviceTests)) to see how it works. 

It's not the same paradigm as Xamarin.UITests (which did everything via automation); sadly, we just don't have a full replacement for that yet.

**Reply by cdefgah** [reply]
@hartez thank you :)
